recipes/cern-root/all/conanfile.py

```diff
@@ -0,0 +1,315 @@
+        "xorg/system",
+        "sqlite3/3.33.0",
+        "libjpeg/9d",
+        "libpng/1.6.37",
+        "libcurl/7.73.0",
+        "pcre/8.44",
+        "xz_utils/5.2.5",
+        "zstd/1.4.5",
+        "lz4/1.9.3",
+        "glew/2.1.0",
+        "openssl/1.1.1i",
+        "fftw/3.3.8",
+        "cfitsio/3.490",
+        "tbb/2020.3",
+        "libpng/1.6.37",
+    )
+
+    _cmake = None
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+
+    @property
+    def _minimum_cpp_standard(self):
+        return 11
+
+    @property
+    def _minimum_compilers_version(self):
+        return {
+            "Visual Studio": "16.1",
+            "gcc": "4.8",
+            "clang": "3.4",
+            "apple-clang": "5.1",
+        }
+
+    def configure(self):
+        if self.settings.compiler.get_safe("cppstd"):
+            tools.check_min_cppstd(self, self._minimum_cpp_standard)
+        min_version = self._minimum_compilers_version.get(str(self.settings.compiler))
+        if not min_version:
+            self.output.warn(
+                "{} recipe lacks information about the {} compiler support.".format(
+                    self.name, self.settings.compiler
+                )
+            )
+        else:
+            if tools.Version(self.settings.compiler.version) < min_version:
+                raise ConanInvalidConfiguration(
+                    "{} requires C++{} support. The current compiler {} {} does not support it.".format(
+                        self.name,
+                        self._minimum_cpp_standard,
+                        self.settings.compiler,
+                        self.settings.compiler.version,
+                    )
+                )
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        os.rename("root-{}".format(self.version.replace("v", "")), self._source_subfolder)
+
+    def _patch_source_cmake(self):
+        os.remove(
+            os.sep.join(
+                (
+                    self._source_subfolder,
+                    "cmake",
+                    "modules",
+                    "FindTBB.cmake",
+                )
+            )
+        )
+        # Conan generated cmake_find_packages names differ from
+        # names ROOT expects (usually only due to case differences)
+        # There is currently no way to change these names
+        # see: https://github.com/conan-io/conan/issues/4430
+        # Patch ROOT CMake to use Conan dependencies
+        tools.replace_in_file(
+            os.path.join(self._source_subfolder, "CMakeLists.txt"),
+            "project(ROOT)",
+            "\n".join(("project(ROOT)",
+            "# sets the current C runtime on MSVC (MT vs MD vd MTd vs MDd)",
+            "include({}/conanbuildinfo.cmake)".format(self.install_folder.replace("\\", "/")),
+            "conan_set_vs_runtime()",
```

Comment:
why not `conan_basic_setup()`?

Honestly I would prefer to add a CMakeLists wrapper like others CCI recipes if it's possible.

Reply:
> why not conan_basic_setup()?

This is being discussed in: <https://github.com/conan-io/conan-center-index/pull/3732#discussion_r555760924>.

In summary: `conan_basic_setup()` was added due to concerns about the C runtime on MSVC. However, this was found to break the build. `conan_set_vs_runtime()` does not cause problems and seems to address the problems that the original commenter had.

At the time of writing, the linked comment thread has not concluded, so it is unclear to me whether this is considered a suitable solution or if some other change is required. I would welcome any feedback that you have on this question.

> Honestly I would prefer to add a CMakeLists.txt wrapper file like others CCI recipes if it's possible.

I agree with you, I would also prefer this. I have tried to implement this following the examples in other CCI recipes. I have tried a wrapper CMakeLists.txt like:

```cmake
cmake_minimum_required(VERSION 3.9)
project(cmake_wrapper)

include(conanbuildinfo)
#conan_basic_setup()
conan_set_vs_runtime()

# Conan generated cmake_find_packages names differ from
# names ROOT expects (usually only due to case differences)
# There is currently no way to change these names
# see: https://github.com/conan-io/conan/issues/4430
find_package(OpenSSL REQUIRED)
set(OPENSSL_VERSION ${OpenSSL_VERSION})
find_package(LibXml2 REQUIRED)
set(LIBXML2_INCLUDE_DIR ${LibXml2_INCLUDE_DIR})
set(LIBXML2_LIBRARIES ${LibXml2_LIBRARIES})
find_package(SQLite3 REQUIRED)
set(SQLITE_INCLUDE_DIR ${SQLITE3_INCLUDE_DIRS})
set(SQLITE_LIBRARIES SQLite::SQLite3)

add_subdirectory("${CMAKE_CURRENT_SOURCE_DIR}/source_subfolder")
```

However, this causes CMake to fail with error messages about missing files.

The issue is that ROOT uses `CMAKE_SOURCE_DIR` in some places within its CMake files in cases where it should `PROJECT_SOURCE_DIR`. When built using add_subdirectory, this variable points to the wrong location and cmake fails.

To fix this we would have to find and patch every instance of this in the ROOT CMake files. But we are now back to patching the CMakeLists.txt so we have the same problem that we were trying to solve.

I am open to alternative suggestions on how to solve this. But if there are none, I think that we need to keep the original solution.

Reply:
Marking this conversation as resolved to keep discussion in one place. 

This is being discussed in the comment thread at: <https://github.com/conan-io/conan-center-index/pull/3732#discussion_r567958665>.